File: src/main/java/io/github/takenoha/towerdefense/tactical/TacticalEffectCompiler.java
package io.github.takenoha.towerdefense.tactical;

import io.github.takenoha.towerdefense.domain.TowerType;
import java.util.EnumMap;
import java.util.EnumSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

/** Compiles selected node snapshots into a constant-time hot-path effect view. */
public final class TacticalEffectCompiler {
    private static final Set<TowerType> ALL_TOWERS =
            Set.copyOf(EnumSet.allOf(TowerType.class));

    public TacticalEffectSnapshot compile(TacticalBuildSelectionView selection) {
        Objects.requireNonNull(selection, "selection");
        Builder builder = new Builder(selection.highestUnlockedTier());
        for (TacticalSkillNodeSnapshot node : selection.nodes()) {
            if (node.tier() <= selection.highestUnlockedTier()) {
                builder.addNode(node);
            }
        }
        return builder.build();
    }

    private static final class Builder {
        private final int highestTier;
        private final EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>> damage =
                conditionalMap();
        private final EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>> attackInterval =
                conditionalMap();
        private final EnumMap<TowerType, Double> rangeAdd = neutralDoubleMap(0.0d);
        private final EnumMap<TowerType, Double> areaRadius = neutralDoubleMap(1.0d);
        private final EnumMap<TowerType, Integer> chainCount = neutralIntegerMap();
        private final EnumMap<TowerType, Double> slowStrength = neutralDoubleMap(1.0d);
        private final EnumMap<TowerType, Double> burnDuration = neutralDoubleMap(1.0d);
        private double supportBuff = 1.0d;
        private double repairCost = 1.0d;
        private double towerDamageTaken = 1.0d;

        private Builder(int highestTier) {
            this.highestTier = highestTier;
        }

        private void addNode(TacticalSkillNodeSnapshot node) {
            if (node.tier() > highestTier) {
                return;
            }
            for (TacticalEffectEntry effect : node.effects()) {
                addEffect(effect);
            }
        }

        private void addEffect(TacticalEffectEntry effect) {
            double value = boundedValue(effect);
            if (!Double.isFinite(value)) {
                return;
            }
            Set<TowerType> targets = effect.towerTypes().isEmpty()
                    ? ALL_TOWERS
                    : effect.towerTypes();
            TacticalTargetCondition condition = derivedCondition(effect);
            switch (effect.type()) {
                case DAMAGE_MULTIPLIER -> addMultiplier(damage, targets, condition, value);
                case ATTACK_INTERVAL_MULTIPLIER ->
                        addMultiplier(attackInterval, targets, condition, value);
                case RANGE_ADD -> addRange(targets, value);
                case AREA_RADIUS_MULTIPLIER -> addPerTowerMultiplier(areaRadius, targets, value);
                case CHAIN_COUNT_ADD -> addChainCount(targets, value);
                case SLOW_STRENGTH_MULTIPLIER ->
                        addPerTowerMultiplier(slowStrength, targets, value);
                case BURN_DURATION_MULTIPLIER ->
                        addPerTowerMultiplier(burnDuration, targets, value);
                case SUPPORT_BUFF_MULTIPLIER -> {
                    if (targets.contains(TowerType.SUPPORT)) {
                        supportBuff = multiply(supportBuff, value);
                    }
                }
                case REPAIR_COST_MULTIPLIER -> {
                    if (targets.equals(ALL_TOWERS)) {
                        repairCost = multiply(repairCost, value);
                    }
                }
                case CORE_LOW_HP_DAMAGE_MULTIPLIER ->
                        addMultiplier(damage, targets, condition, value);
                case CORE_LOW_HP_ATTACK_INTERVAL_MULTIPLIER ->
                        addMultiplier(attackInterval, targets, condition, value);
                case TOWER_DAMAGE_TAKEN_MULTIPLIER -> {
                    if (targets.equals(ALL_TOWERS)) {
                        towerDamageTaken = multiply(towerDamageTaken, value);
                    }
                }
                case DAMAGE_TO_BOSS_MULTIPLIER,
                        DAMAGE_TO_HIGH_HP_MULTIPLIER,
                        DAMAGE_TO_LOW_HP_MULTIPLIER,
                        DAMAGE_TO_SLOWED_TARGET_MULTIPLIER,
                        DAMAGE_TO_BURNING_TARGET_MULTIPLIER ->
                        addMultiplier(damage, targets, condition, value);
            }
        }

        private TacticalEffectSnapshot build() {
            return new TacticalEffectSnapshotImpl(
                    damage,
                    attackInterval,
                    rangeAdd,
                    areaRadius,
                    chainCount,
                    slowStrength,
                    burnDuration,
                    supportBuff,
                    repairCost,
                    towerDamageTaken);
        }

        private static void addMultiplier(
                EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>> values,
                Set<TowerType> targets,
                TacticalTargetCondition condition,
                double value) {
            if (value <= 0.0d) {
                return;
            }
            for (TowerType type : targets) {
                EnumMap<TacticalTargetCondition, Double> row = values.get(type);
                row.put(condition, multiply(row.getOrDefault(condition, 1.0d), value));
            }
        }

        private void addRange(Set<TowerType> targets, double value) {
            if (value < 0.0d) {
                return;
            }
            for (TowerType type : targets) {
                double previous = rangeAdd.getOrDefault(type, 0.0d);
                rangeAdd.put(type, previous + value);
            }
        }

        private static void addPerTowerMultiplier(
                EnumMap<TowerType, Double> values,
                Set<TowerType> targets,
                double value) {
            if (value <= 0.0d) {
                return;
            }
            for (TowerType type : targets) {
                values.put(type, multiply(values.getOrDefault(type, 1.0d), value));
            }
        }

        private void addChainCount(Set<TowerType> targets, double value) {
            if (value < 0.0d || value > Integer.MAX_VALUE
                    || value != Math.rint(value)) {
                return;
            }
            int delta = (int) value;
            for (TowerType type : targets) {
                long sum = (long) chainCount.getOrDefault(type, 0) + delta;
                chainCount.put(type, sum > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) sum);
            }
        }

        private static double boundedValue(TacticalEffectEntry effect) {
            double value = effect.value();
            if (effect.minimum() != null) {
                value = Math.max(value, effect.minimum());
            }
            if (effect.maximum() != null) {
                value = Math.min(value, effect.maximum());
            }
            return value;
        }

        private static TacticalTargetCondition derivedCondition(TacticalEffectEntry effect) {
            if (effect.condition() != TacticalTargetCondition.NONE) {
                return effect.condition();
            }
            return switch (effect.type()) {
                case CORE_LOW_HP_DAMAGE_MULTIPLIER -> TacticalTargetCondition.CORE_BELOW_50_PERCENT;
                case CORE_LOW_HP_ATTACK_INTERVAL_MULTIPLIER ->
                        TacticalTargetCondition.CORE_BELOW_30_PERCENT;
                case DAMAGE_TO_BOSS_MULTIPLIER -> TacticalTargetCondition.BOSS;
                case DAMAGE_TO_HIGH_HP_MULTIPLIER -> TacticalTargetCondition.HIGH_HP;
                case DAMAGE_TO_LOW_HP_MULTIPLIER -> TacticalTargetCondition.LOW_HP;
                case DAMAGE_TO_SLOWED_TARGET_MULTIPLIER -> TacticalTargetCondition.SLOWED;
                case DAMAGE_TO_BURNING_TARGET_MULTIPLIER -> TacticalTargetCondition.BURNING;
                default -> TacticalTargetCondition.NONE;
            };
        }

        private static double multiply(double left, double right) {
            return TacticalEffectSnapshotImpl.safeProduct(left, right);
        }

        private static EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>> conditionalMap() {
            EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>> map =
                    new EnumMap<>(TowerType.class);
            for (TowerType type : TowerType.values()) {
                map.put(type, new EnumMap<>(TacticalTargetCondition.class));
            }
            return map;
        }

        private static EnumMap<TowerType, Double> neutralDoubleMap(double value) {
            EnumMap<TowerType, Double> map = new EnumMap<>(TowerType.class);
            for (TowerType type : TowerType.values()) {
                map.put(type, value);
            }
            return map;
        }

        private static EnumMap<TowerType, Integer> neutralIntegerMap() {
            EnumMap<TowerType, Integer> map = new EnumMap<>(TowerType.class);
            for (TowerType type : TowerType.values()) {
                map.put(type, 0);
            }
            return map;
        }
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/tactical/TacticalEffectCompiler.kt
package io.github.takenoha.towerdefense.tactical

import io.github.takenoha.towerdefense.domain.TowerType
import java.util.EnumMap
import java.util.EnumSet
import java.util.Objects

private val ALL_TOWERS: Set<TowerType> =
    java.util.Set.copyOf(EnumSet.allOf(TowerType::class.java)).toSet()

/** Compiles selected node snapshots into a constant-time hot-path effect view. */
class TacticalEffectCompiler {
    fun compile(selection: TacticalBuildSelectionView): TacticalEffectSnapshot {
        Objects.requireNonNull(selection, "selection")
        val builder = Builder(selection.highestUnlockedTier())
        for (node in selection.nodes()) {
            val unlocked = if (selection.unlockedNodeIds().isEmpty()) {
                node.tier() <= selection.highestUnlockedTier()
            } else {
                selection.unlockedNodeIds().contains(node.id())
            }
            if (unlocked) {
                builder.addNode(node)
            }
        }
        return builder.build()
    }

    private class Builder(private val highestTier: Int) {
        private val damage = conditionalMap()
        private val attackInterval = conditionalMap()
        private val rangeAdd = neutralDoubleMap(0.0)
        private val areaRadius = neutralDoubleMap(1.0)
        private val chainCount = neutralIntegerMap()
        private val slowStrength = neutralDoubleMap(1.0)
        private val burnDuration = neutralDoubleMap(1.0)
        private var supportBuff = 1.0
        private var repairCost = 1.0
        private var towerDamageTaken = 1.0

        fun addNode(node: TacticalSkillNodeSnapshot) {
            if (node.tier() > highestTier) {
                return
            }
            for (effect in node.effects()) {
                addEffect(effect)
            }
        }

        private fun addEffect(effect: TacticalEffectEntry) {
            val value = boundedValue(effect)
            if (!value.isFinite()) {
                return
            }
            val targets: Set<TowerType> = if (effect.towerTypes().isEmpty()) {
                ALL_TOWERS
            } else {
                effect.towerTypes().toSet()
            }
            val condition = derivedCondition(effect)
            when (effect.type()) {
                TacticalEffectType.DAMAGE_MULTIPLIER ->
                    addMultiplier(damage, targets, condition, value)
                TacticalEffectType.ATTACK_INTERVAL_MULTIPLIER ->
                    addMultiplier(attackInterval, targets, condition, value)
                TacticalEffectType.RANGE_ADD -> addRange(targets, value)
                TacticalEffectType.AREA_RADIUS_MULTIPLIER ->
                    addPerTowerMultiplier(areaRadius, targets, value)
                TacticalEffectType.CHAIN_COUNT_ADD -> addChainCount(targets, value)
                TacticalEffectType.SLOW_STRENGTH_MULTIPLIER ->
                    addPerTowerMultiplier(slowStrength, targets, value)
                TacticalEffectType.BURN_DURATION_MULTIPLIER ->
                    addPerTowerMultiplier(burnDuration, targets, value)
                TacticalEffectType.SUPPORT_BUFF_MULTIPLIER -> {
                    if (targets.contains(TowerType.SUPPORT)) {
                        supportBuff = multiply(supportBuff, value)
                    }
                }
                TacticalEffectType.REPAIR_COST_MULTIPLIER -> {
                    if (targets == ALL_TOWERS) {
                        repairCost = multiply(repairCost, value)
                    }
                }
                TacticalEffectType.CORE_LOW_HP_DAMAGE_MULTIPLIER ->
                    addMultiplier(damage, targets, condition, value)
                TacticalEffectType.CORE_LOW_HP_ATTACK_INTERVAL_MULTIPLIER ->
                    addMultiplier(attackInterval, targets, condition, value)
                TacticalEffectType.TOWER_DAMAGE_TAKEN_MULTIPLIER -> {
                    if (targets == ALL_TOWERS) {
                        towerDamageTaken = multiply(towerDamageTaken, value)
                    }
                }
                TacticalEffectType.DAMAGE_TO_BOSS_MULTIPLIER,
                TacticalEffectType.DAMAGE_TO_HIGH_HP_MULTIPLIER,
                TacticalEffectType.DAMAGE_TO_LOW_HP_MULTIPLIER,
                TacticalEffectType.DAMAGE_TO_SLOWED_TARGET_MULTIPLIER,
                TacticalEffectType.DAMAGE_TO_BURNING_TARGET_MULTIPLIER ->
                    addMultiplier(damage, targets, condition, value)
            }
        }

        fun build(): TacticalEffectSnapshot = TacticalEffectSnapshotImpl(
            damage,
            attackInterval,
            rangeAdd,
            areaRadius,
            chainCount,
            slowStrength,
            burnDuration,
            supportBuff,
            repairCost,
            towerDamageTaken,
        )

        private fun addMultiplier(
            values: EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>>,
            targets: Set<TowerType>,
            condition: TacticalTargetCondition,
            value: Double,
        ) {
            if (value <= 0.0) {
                return
            }
            for (type in targets) {
                val row = values[type]!!
                row[condition] = multiply(row.getOrDefault(condition, 1.0), value)
            }
        }

        private fun addRange(targets: Set<TowerType>, value: Double) {
            if (value < 0.0) {
                return
            }
            for (type in targets) {
                val previous = rangeAdd.getOrDefault(type, 0.0)
                rangeAdd[type] = previous + value
            }
        }

        private fun addPerTowerMultiplier(
            values: EnumMap<TowerType, Double>,
            targets: Set<TowerType>,
            value: Double,
        ) {
            if (value <= 0.0) {
                return
            }
            for (type in targets) {
                values[type] = multiply(values.getOrDefault(type, 1.0), value)
            }
        }

        private fun addChainCount(targets: Set<TowerType>, value: Double) {
            if (value < 0.0 || value > Int.MAX_VALUE || value != Math.rint(value)) {
                return
            }
            val delta = value.toInt()
            for (type in targets) {
                val sum = chainCount.getOrDefault(type, 0).toLong() + delta.toLong()
                chainCount[type] = if (sum > Int.MAX_VALUE) Int.MAX_VALUE else sum.toInt()
            }
        }

        private fun boundedValue(effect: TacticalEffectEntry): Double {
            var value = effect.value()
            val minimum = effect.minimum()
            if (minimum != null) {
                value = maxOf(value, minimum)
            }
            val maximum = effect.maximum()
            if (maximum != null) {
                value = minOf(value, maximum)
            }
            return value
        }

        private fun derivedCondition(effect: TacticalEffectEntry): TacticalTargetCondition {
            if (effect.condition() != TacticalTargetCondition.NONE) {
                return effect.condition()
            }
            return when (effect.type()) {
                TacticalEffectType.CORE_LOW_HP_DAMAGE_MULTIPLIER ->
                    TacticalTargetCondition.CORE_BELOW_50_PERCENT
                TacticalEffectType.CORE_LOW_HP_ATTACK_INTERVAL_MULTIPLIER ->
                    TacticalTargetCondition.CORE_BELOW_30_PERCENT
                TacticalEffectType.DAMAGE_TO_BOSS_MULTIPLIER -> TacticalTargetCondition.BOSS
                TacticalEffectType.DAMAGE_TO_HIGH_HP_MULTIPLIER -> TacticalTargetCondition.HIGH_HP
                TacticalEffectType.DAMAGE_TO_LOW_HP_MULTIPLIER -> TacticalTargetCondition.LOW_HP
                TacticalEffectType.DAMAGE_TO_SLOWED_TARGET_MULTIPLIER -> TacticalTargetCondition.SLOWED
                TacticalEffectType.DAMAGE_TO_BURNING_TARGET_MULTIPLIER ->
                    TacticalTargetCondition.BURNING
                else -> TacticalTargetCondition.NONE
            }
        }

        private fun multiply(left: Double, right: Double): Double =
            TacticalEffectSnapshotImpl.safeProduct(left, right)

        private fun conditionalMap(): EnumMap<
            TowerType,
            EnumMap<TacticalTargetCondition, Double>,
        > {
            val map = EnumMap<TowerType, EnumMap<TacticalTargetCondition, Double>>(
                TowerType::class.java,
            )
            for (type in TowerType.values()) {
                map[type] = EnumMap(TacticalTargetCondition::class.java)
            }
            return map
        }

        private fun neutralDoubleMap(value: Double): EnumMap<TowerType, Double> {
            val map = EnumMap<TowerType, Double>(TowerType::class.java)
            for (type in TowerType.values()) {
                map[type] = value
            }
            return map
        }

        private fun neutralIntegerMap(): EnumMap<TowerType, Int> {
            val map = EnumMap<TowerType, Int>(TowerType::class.java)
            for (type in TowerType.values()) {
                map[type] = 0
            }
            return map
        }
    }
}
